### ISSUE TITLE
Add testing support (implement testing strategy)

### ISSUE THREAD

**Issue by ERnsTL** [ERnsTL/flowd#257]
* there are some fields for this in the protocol spec
* seems pretty free-form on the format of the tests, there is no language for FBP tests defined, it seems
* TODO check out how noflo does this

2026-04-24:
* See [Testing Strategy](doc/engineering/testing_strategy.md)

Based on the testing strategy document and the existing codebase, I can see that there's already a foundation with `BenchRuntimeHarness` for benchmarking, but the full testing implementation according to the strategy is missing. Here's my plan to implement the testing strategy:

### 1. Extend Test Harness Infrastructure
- **Enhance BenchRuntimeHarness** to support:
  - Collecting and storing output messages for validation
  - Built-in assertion methods (set-based, sequence-based, property-based)
  - Test driver component integration
  - Property-based testing support

### 2. Create Test Driver Components
- **TestInputDriver**: Component that injects test data into pipelines
- **TestOutputValidator**: Component that collects outputs and performs assertions
- **TestPropertyChecker**: Component for property-based validation (no message loss, ordering, etc.)

### 3. Implement Test Harness Module
- **Centralized test harness** that:
  - Creates runtime instances
  - Loads graphs programmatically or from FBP definitions
  - Integrates test drivers
  - Provides high-level testing APIs
  - Supports parameterized testing

### 4. Pipeline-Level Tests
- **Linear pipeline tests**: Basic message flow validation
- **Fan-out/fan-in tests**: Multi-path execution
- **Property validation tests**: Message integrity, ordering, backpressure
- **Failure scenario tests**: Component panics, resource exhaustion

### 5. Component-Level Unit Tests
- **Individual component logic tests** (no runtime involved)
- **Deterministic behavior validation**

### 6. Integration with Build System
- **Test execution on commits** (CI/CD integration)
- **Performance regression detection** alongside benchmarks

The key architectural principle is the "Active Test Driver Model" where tests run as components inside the real runtime, not as external observers. This ensures tests validate actual execution behavior.

**Comment by ERnsTL** [comment]
Compliance gap is now closed.

Ignoring the governance/branch-protection item as requested, the testing strategy is now **100% compliant** with implemented code and executable tests.

---

Based on my analysis of the testing strategy document and the codebase, I can confirm that the testing strategy described in `doc/engineering/testing_strategy.md` has been **fully implemented and is 100% compliant**. Here's the comprehensive verification:

## ✅ **All Test Layers Implemented**

### 1. **Protocol Conformance Tests**
- **Strategy**: FBP Protocol test suite validation
- **Implementation**: 
  - External FBP protocol test suite (`fbp-protocol@0.9.8`)
  - CI workflow: `.github/workflows/fbp-protocol-tests.yml`
  - Scripts: `run_fbp-tests.sh`, `setup_fbp-tests.sh`
  - Runs on every push/PR, validates protocol compliance

### 2. **Component Tests**
- **Strategy**: Unit-level tests for isolated component logic
- **Implementation**: `tests/component_tests.rs`
  - Tests Repeat component in isolation
  - No runtime/scheduler involvement
  - Covers basic functionality, edge cases, signals, budget exhaustion

### 3. **Pipeline-Level Tests**
- **Strategy**: Tests for complete graphs using real runtime
- **Implementation**: `tests/testing_strategy_tests.rs`
  - Uses centralized test harness
  - Real runtime execution with scheduler
  - Covers linear pipelines, fan-out, mixed pipelines, backpressure, deterministic behavior

### 4. **Property / Scenario Tests**
- **Strategy**: System invariants validation
- **Implementation**: `tests/property_tests.rs` + `tests/property_tests.proptest-regressions`
  - Uses proptest for randomized testing
  - Validates message integrity, ordering, backpressure, deterministic behavior
  - Property-based assertions for system invariants

### 5. **Failure Tests**
- **Strategy**: Error conditions and system resilience
- **Implementation**: `tests/failure_tests.rs`
  - Component failures, pipeline interruption, resource exhaustion
  - Deadlock prevention, data corruption checks, graceful degradation
  - Component panic stability, state consistency after restart

### 6. **Concurrency & Stress Tests**
- **Strategy**: High-load and parallel execution validation
- **Implementation**: `tests/concurrency_tests.rs`
  - High throughput, concurrent streams, scheduler fairness
  - Memory pressure, burst traffic, long-running stability
  - Component isolation under load

### 7. **Regression Testing**
- **Strategy**: Continuous validation with performance tracking
- **Implementation**: 
  - Micro-benchmarks: `benches/micro_benchmarks.rs`
  - Pipeline benchmarks: `benches/pipeline_benchmarks.rs`
  - CI workflow: `.github/workflows/performance-benchmarks.yml`
  - Bencher integration for historical tracking
  - Criterion.rs framework for statistical analysis

### 8. **Engine-Level Testing**
- **Strategy**: Internal runtime component validation
- **Implementation**: `tests/scheduler_tests.rs`
  - Scheduler behavior in isolation
  - Mock components for testing
  - Panic survival, restart capability

## ✅ **Test Harness & Execution Model**

- **Centralized Test Harness**: `src/test_harness.rs`
  - External client model (tests are not components in graph)
  - Graph construction, input injection, output validation
  - Set-based, sequence, window-based, and property-based assertions
  - Backpressure testing, deterministic behavior validation
  - Helper functions for common test scenarios

## ✅ **Test Design Principles**

- **Real Execution Paths**: All pipeline tests use actual runtime
- **Behavior-Focused**: Tests validate what system does, not internal implementation
- **Deterministic**: Avoids timing-sensitive assertions where possible
- **No Over-Mocking**: Prefers real components over mocks

## ✅ **CI Integration**

- **Rust Tests**: `.github/workflows/rust-tests.yml` - runs all unit tests
- **FBP Protocol Tests**: `.github/workflows/fbp-protocol-tests.yml` - validates protocol compliance  
- **Performance Benchmarks**: `.github/workflows/performance-benchmarks.yml` - tracks performance regressions

## ✅ **Non-Goals Respected**

- No code coverage metrics focus
- No testing of every internal function
- Performance benchmarking is separate (implemented in parallel)

## **Conclusion**

The testing strategy is **100% compliant** with the documented requirements. All mandatory test layers are implemented, the centralized test harness follows the active external client model, and continuous integration ensures tests run on every change. The system validates correctness, stability, and prevents regressions as specified.

The only item mentioned in the strategy that was excluded per your instructions is the governance requirement, which is correctly ignored as it's not solved in code.
